File: App/src/Adaptive.ReactiveTrader.Client.Domain/Models/ReferenceData/CurrencyPair.cs
using System;
using System.Reactive.Linq;
using Adaptive.ReactiveTrader.Client.Domain.Models.Pricing;
using Adaptive.ReactiveTrader.Client.Domain.Repositories;

namespace Adaptive.ReactiveTrader.Client.Domain.Models.ReferenceData
{
    internal class CurrencyPair : ICurrencyPair
    {
        private readonly Lazy<IObservable<IPrice>> _lazyPrices;

        public CurrencyPair(string symbol, int ratePrecision, int pipsPosition, IPriceRepository priceRepository)
        {
            Symbol = symbol;
            RatePrecision = ratePrecision;
            PipsPosition = pipsPosition;
            BaseCurrency = symbol.Substring(0, 3);
            CounterCurrency = symbol.Substring(3, 3);

            _lazyPrices = new Lazy<IObservable<IPrice>>(() => CreatePrices(priceRepository));
        }


        private IObservable<IPrice> CreatePrices(IPriceRepository priceRepository)
        {
            return priceRepository.GetPrices(this)
                .Publish()
                .RefCount();
        }

        public string Symbol { get; private set; }
        public int RatePrecision { get; private set; }
        public int PipsPosition { get; private set; }
        public string BaseCurrency { get; private set; }
        public string CounterCurrency { get; private set; }
        public IObservable<IPrice> Prices { get { return _lazyPrices.Value; } }
    }
}

File: App/src/Adaptive.ReactiveTrader.Client.JS/Models/ReferenceData/CurrencyPair.ts
class CurrencyPair implements ICurrencyPair {
    private _symbol: string;
    private _ratePrecision: number;
    private _pipsPosition: number;
    private _baseCurrency: string;
    private _counterCurrency: string;
    private _prices: Rx.Observable<IPrice>

    constructor(symbol: string, ratePrecision: number, pipsPosition: number, priceRespository: IPriceRepository) {
        this._symbol = symbol;
        this._ratePrecision = ratePrecision;
        this._pipsPosition = pipsPosition;
        this._baseCurrency = symbol.substring(0, 3);
        this._counterCurrency = symbol.substring(3, 6);

        this._prices =
            Rx.Observable.defer<IPrice>(()=> priceRespository.getPrices(this))
            .publish()
            .refCount();
    }

    public get Symbol() {
        return this._symbol;
    }

    public get RatePrecision() {
        return this._ratePrecision;
    }

    public get PipsPositio() {
        return this._pipsPosition;
    }

    public get BaseCurrency() {
        return this._baseCurrency;
    }

    public get CounterCurrency() {
        return this._counterCurrency;
    }

    public get Prices() {
        return this._prices;
    }
} 